src/IceRpc/Transports/Slic/Internal/SlicDuplexConnectionDecorator.cs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) ZeroC, Inc.
+
+using System.Buffers;
+using System.Diagnostics;
+
+namespace IceRpc.Transports.Slic.Internal;
+
+/// <summary>Decorates <see cref="ReadAsync" /> to fail if no byte is received for over idle timeout. Also optionally
+/// decorates both <see cref="ReadAsync"/> and <see cref="WriteAsync" /> to schedule pings that prevent both the local
+/// and remote idle timers from expiring.</summary>
+internal class SlicDuplexConnectionDecorator : IDuplexConnection
+{
+    private readonly IDuplexConnection _decoratee;
+    private TimeSpan _idleTimeout = Timeout.InfiniteTimeSpan;
+    private readonly CancellationTokenSource _readCts = new();
+
+    private Timer? _readTimer;
+    private Timer? _writeTimer;
+
+    public Task<TransportConnectionInformation> ConnectAsync(CancellationToken cancellationToken) =>
+        _decoratee.ConnectAsync(cancellationToken);
+
+    public void Dispose()
+    {
+        _decoratee.Dispose();
+        _readCts.Dispose();
+
+        // Using Dispose is fine, there's no need to wait for the keep alive action to terminate if it's running.
+        _readTimer?.Dispose();
+        _writeTimer?.Dispose();
+    }
+
+    public ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+    {
+        return _idleTimeout == Timeout.InfiniteTimeSpan ?
+            _decoratee.ReadAsync(buffer, cancellationToken) :
+            PerformReadAsync();
+
+        async ValueTask<int> PerformReadAsync()
+        {
+            try
+            {
+                using CancellationTokenRegistration _ = cancellationToken.UnsafeRegister(
+                    cts => ((CancellationTokenSource)cts!).Cancel(),
+                    _readCts);
+                _readCts.CancelAfter(_idleTimeout); // enable idle timeout before reading
+
+                int bytesRead = await _decoratee.ReadAsync(buffer, _readCts.Token).ConfigureAwait(false);
+                // Debug.Assert(bytesRead > 0); // TODO: uncomment when bug is fixed
```

Comment:
What bug? It would be nicer to link to a GitHub issue.
```suggestion
                // Debug.Assert(bytesRead > 0); // TODO: uncomment when bug is fixed
```

Reply:
Fixed